In [2]:
from week1 import calculateDatasetShannonEntropy
calculateDatasetShannonEntropy([0,0,1,1])

1.0

# samples from a normal distribution with mu  and sigma using numpy
# For random samples from N(\mu, \sigma^2), use:

# sigma * np.random.randn(...) + mu

numpy.random.randint(low, high=None, size=None, dtype='l')
Return random integers from low (inclusive) to high (exclusive).

Return random integers from the “discrete uniform” distribution of the specified dtype in the “half-open” interval [low, high). If high is None (the default), then results are from [0, low).

In [4]:
import numpy as np
samples = np.random.randint(0,2, 10)
print(samples)

[1 0 1 0 0 1 1 0 0 1]


In [9]:
from week1 import entropy1, calculateDatasetShannonEntropy
Hs = calculateDatasetShannonEntropy(samples)
print(Hs)

1.0


In [14]:
samples = np.random.randint(0,4, 100)
print(samples)
Hs = calculateDatasetShannonEntropy(samples)
print(Hs)

[3 2 3 0 1 0 2 2 1 0 3 3 0 0 3 0 2 2 3 3 2 1 1 0 1 1 1 0 2 3 0 2 1 0 0 3 1
 1 1 2 0 0 0 0 0 0 2 1 2 3 3 2 0 3 0 2 2 0 2 3 2 1 1 0 2 2 1 3 2 2 1 0 2 0
 3 3 1 2 3 0 2 3 1 3 0 3 3 3 1 0 0 2 2 2 0 3 2 2 1 2]
1.9841774541410144


In [78]:
from math import log2, isclose
def entropy1(p: np.array):
    if type(p) == list:
        p = np.array(p)
    # if np.sum(p) != 1:
    if not isclose(np.sum(p), 1, rel_tol=1e-):
        raise Exception('The sum of the elements of p should be = 1: {}'.format(p))
    with suppress(ZeroDivisionError):
        H = - np.dot(p, np.where(p > 0, np.log2(p), 0))

    return  H

# Conditional entropy

In [79]:
def conditionalentropy(p: np.array):
    # entropyXGivenY
    # H(X|Y)
    # 2 dim only for now

    # e.g.
    # p(x,y)
    # coded for Y rows (dim=0) \X   columns (dim=1)
    # =>
    # transpose(p)
    # p = np.array([[0.125, 0.0625, 0.03125, 0.03125],
    #       [0.0625, 0.125, 0.03125, 0.03125],
    #       [0.0625, 0.0625, 0.0625, 0.0625],
    #       [0.25,0,0,0]])
    # Correct answer = 11/8 = 1.375 fro transpose p                    
    if (type(p) == list):
        p = np.array(p)

    p = np.transpose(p)

    total = 0
    for i in range(p.shape[0]):
        pyi = np.sum(p[i])
        total+= pyi * entropy1(1/pyi * p[i])

    return total

In [76]:
p1 = np.array([[0.2, 0.3],[0.1, 0.4]])
print(p1)
print(p1[:,0])

[[0.2 0.3]
 [0.1 0.4]]
[0.2 0.1]


In [87]:
def conditionalentropy2(p: np.array):
    # entropyXGivenY
    # H(X|Y)
    # 2 dim only for now

       # X = rows
     # Y = columnscoded
    if (type(p) == list):
        p = np.array(p)

    total = 0
    for i in range(p.shape[1]):
        pxi = np.sum(p[i])
        total+= :,pxi * entropy1(1/pxi * p[i])

    return total:,

In [88]:
print(conditionalentropy(p1))
print(conditionalentropy2(p1))

0.9651484454403229
0.9651484454403229


# Joint entropy

In [31]:
from week1 import jointEntropy


In [106]:
p2 = np.array([[0.2, 0.3],[0.1, 0.4]])
jointEntropy(p2)

1.8464393446710154

In [25]:
np.sum(p2)

1.0

In [96]:
from math import log2, isclose
import numpy as np
import matplotlib.pyplot as plt
from contextlib import suppress
import collections
import scipy.stats as scips

#test:p = np.array([[0.2, 0.3],[0.1, 0.4]])
# => jointEntropy : 1.8464393446710154

def jointEntropy(p: np.array):

    if type(p) == list:
        p = np.array(p)

    if ( np.sum(p) != 1) :
        raise Exception('The sum of the elements of p should be = 1{}'.format(p))

    # Perform element-wise p(x,y)*log p(x,y)
    plogp = lambda x: - x * np.where(x > 0, np.log2(x), 0)
    Hs = np.array([plogp(pij) for pij in p])

    return  np.sum(Hs)

In [97]:
p2 = np.array([[0.2, 0.3],[0.1, 0.4]])
jointEntropy(p2)

1.8464393446710154

In [43]:
plogp = lambda p: - p * np.where(p > 0, np.log2(p), 0)


In [44]:
plogp(0.34)

0.5291737384982911

In [45]:
plogp(0)

<ipython-input-43-f8d1d2e4c4c3>:1: RuntimeWarning: divide by zero encountered in log2
  plogp = lambda p: - p * np.where(p > 0, np.log2(p), 0)


0.0

In [49]:
p = np.array([[0.2, 0.3],[0.1, 0.4]])
Hs = np.array([plogp(pij) for pij in p])
np.sum(Hs)

1.8464393446710154

# joint entropy in more than 2 dimensions

In [109]:

def jointEntropy(p: np.array):

    if type(p) == list:
        p = np.array(p)

    if not isclose(np.sum(p), 1, rel_tol=1e-6):
        raise Exception('The sum of the elements of p should be = 1{}'.format(p))

    # Perform element-wise p(x,y)*log p(x,y)
    plogp = lambda x: - x * np.where(x > 0, np.log2(x), 0)
    Hs = np.array([plogp(pij) for pij in p])

    return  np.sum(Hs)

p3 = np.zeros((3,3,3))
p3[0,0,0] = 1
print(p3)
np.sum(p3)

[[[1. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]]


1.0

In [110]:
jointEntropy(p3)

<ipython-input-109-b56250aecc45>:10: RuntimeWarning: divide by zero encountered in log2
  plogp = lambda x: - x * np.where(x > 0, np.log2(x), 0)


0.0

In [111]:
p4 = np.array([1/27 for i in range(27)]).reshape((3,3,3))
print(p4)

[[[0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]]

 [[0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]]

 [[0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]]]


In [112]:
jointEntropy(p4)

4.754887502163469

In [102]:
np.sum(p4)

0.9999999999999998

# Coding Joint entropy

In [ ]:
def jointEntropy(p: np.array):

    if type(p) == list:
        p = np.array(p)

    if not isclose(np.sum(p), 1, rel_tol=1e-6):
        raise Exception('The sum of the elements of p should be = 1{}'.format(p))

    # Perform element-wise p(x,y)*log p(x,y)
    plogp = lambda x: - x * np.where(x > 0, np.log2(x), 0)
    Hs = np.array([plogp(pij) for pij in p])

    return  np.sum(Hs)

Test that your code works, e.g. by running jointentropy([0.2, 0.3; 0.1, 0.4]) and validating that you get the result 1.85 bits

In [115]:
jointEntropy([[0.2, 0.3],[ 0.1, 0.4]])

1.8464393446710154

## joint entropy in multiple dimensions

In [116]:
p3 = np.zeros((3,3,3))
p3[0,0,0] = 1
print(p3)
np.sum(p3)

[[[1. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 0.]]]


1.0

In [119]:
jointEntropy(p3)

<ipython-input-109-b56250aecc45>:10: RuntimeWarning: divide by zero encountered in log2
  plogp = lambda x: - x * np.where(x > 0, np.log2(x), 0)


0.0

In [117]:
p4 = np.array([1/27 for i in range(27)]).reshape((3,3,3))
print(p4)

[[[0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]]

 [[0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]]

 [[0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]
  [0.03703704 0.03703704 0.03703704]]]


In [118]:
jointEntropy(p4)

4.754887502163469

# 8. (Extension) Coding joint entropy for empirical data

rows = samples x[i] of X (i=1..N, number of samples)
columns = variables/features  
here will always be multiple rows (because we always should have many samples), but the number of columns will depend on how many variables we are jointly considering (it could be just one if we have only a single variable).

In [124]:
import itertools
from collections import Counter

some_list = [['foobar', 'acling', 'b'], ['xbar', 'cling', 'd'], ['foobar', 'dcling', 'b'], ['xbar', 'acling', 'd']]
# totals = Counter(i for i in list(itertools.chain.from_iterable(some_list)))

some_list = np.array(some_list)
totals2 = Counter(list(some_list.flatten()))

print(some_list)
print(some_list.flatten())
print(list(some_list.flatten()))
print(totals2)

def jointEntropyEmpirical(samples: np.array):

    # check only 2D

    # Each row of item represent a sample with the columns being the random variables or features of interest
    # row 0 : X
    # row 1 : Y

    H, edges = np.histogramdd(samples)

    for i in range(p.shape[0]):  #iterate over samples
        xiCount = collections.Counter(items[i])


def calculateDatasetShannonEntropy(items):
    # also called empiricalentropy in course
    itemscount = collections.Counter(items)
    
    # probability = # item x / # total number of items
    dist = [x/sum(itemscount.values()) for x in itemscount.values()]
 
    # use scipy to calculate entropy
    entropy_value = scips.entropy(dist, base=2)
 
    return entropy_value

[['foobar' 'acling' 'b']
 ['xbar' 'cling' 'd']
 ['foobar' 'dcling' 'b']
 ['xbar' 'acling' 'd']]
['foobar' 'acling' 'b' 'xbar' 'cling' 'd' 'foobar' 'dcling' 'b' 'xbar'
 'acling' 'd']
['foobar', 'acling', 'b', 'xbar', 'cling', 'd', 'foobar', 'dcling', 'b', 'xbar', 'acling', 'd']
Counter({'foobar': 2, 'acling': 2, 'b': 2, 'xbar': 2, 'd': 2, 'cling': 1, 'dcling': 1})


https://stackoverflow.com/questions/20332750/python-joint-distribution-of-n-variables    
https://numpy.org/doc/stable/reference/generated/numpy.histogram2d.html#numpy.histogram2d

In [136]:
samples = [['foobar', 'cling'], ['xbar', 'cling'], ['foobar', 'dcling'], ['xbar', 'acling']]
# totals = Counter(i for i in list(itertools.chain.from_iterable(some_list)))
samples = np.array(samples)

alphabetX=list(set(samples[:,0]))
print(alphabetX)
print(len(alphabetX))

alphabetY=list(set(samples[:,1]))
print(alphabetY)
print(len(alphabetY))

['foobar', 'xbar']
2
['dcling', 'cling', 'acling']
3


In [164]:

alphabetX=list(set(samples[:,0]))
alphabetY=list(set(samples[:,1]))

jointProbabilities=np.zeros((len(alphabetX), len(alphabetY)))

for i in range(samples.shape[0]):
    print('Sample {} : {}'.format(i, samples[i]))
    sample=samples[i]
    jointProbabilities[alphabetX.index(sample[0]), alphabetY.index(sample[1])]+=1

jointProbabilities/=samples.shape[0]

jointEntropy(jointProbabilities)


Sample 0 : ['foobar' 'cling']
Sample 1 : ['xbar' 'cling']
Sample 2 : ['foobar' 'dcling']
Sample 3 : ['xbar' 'acling']


In [165]:
print(alphabetX)
print(alphabetY)
print(jointProbabilities)

['foobar', 'xbar']
['dcling', 'cling', 'acling']
[[0.25 0.25 0.  ]
 [0.   0.25 0.25]]


In [158]:
jointEntropy(jointProbabilities)

<ipython-input-109-b56250aecc45>:10: RuntimeWarning: divide by zero encountered in log2
  plogp = lambda x: - x * np.where(x > 0, np.log2(x), 0)


2.0